src/emulate.c

```diff
@@ -1884,171 +1823,44 @@ void rv_debug(struct riscv_t *rv)
-        OP(cfld),      OP(cjal),      OP(cfldsp), OP(unimp),  // 001
-        OP(clw),       OP(cli),       OP(clwsp),  OP(unimp),  // 010
-        OP(cflw),      OP(clui),      OP(cflwsp), OP(unimp),  // 011
-        OP(unimp),     OP(cmisc_alu), OP(ccr),    OP(unimp),  // 100
-        OP(cfsd),      OP(cj),        OP(cfsdsp), OP(unimp),  // 101
-        OP(csw),       OP(cbeqz),     OP(cswsp),  OP(unimp),  // 110
-        OP(cfsw),      OP(cbnez),     OP(cfswsp), OP(unimp),  // 111
-    };
-#endif
-    /* clang-format on */
-
-#if RV32_HAS(COMPUTED_GOTO)
-#if RV32_HAS(EXT_C)
-#define DISPATCH_RV32C()                                            \
-    insn &= 0x0000FFFF;                                             \
-    int16_t c_index = (insn & FC_FUNC3) >> 11 | (insn & FC_OPCODE); \
-    rv->insn_len = INSN_16;                                         \
-    goto *jump_table_rvc[c_index];
-#else
-#define DISPATCH_RV32C()
-#endif
-
-#define DISPATCH()                                                \
-    {                                                             \
-        if (unlikely(rv->csr_cycle >= cycles_target || rv->halt)) \
-            return;                                               \
-        /* fetch the next instruction */                          \
-        insn = rv->io.mem_ifetch(rv, rv->PC);                     \
-        /* standard uncompressed instruction */                   \
-        if ((insn & 3) == 3) {                                    \
-            uint32_t index = (insn & INSN_6_2) >> 2;              \
-            rv->insn_len = INSN_32;                               \
-            goto *jump_table[index];                              \
-        } else {                                                  \
-            /* Compressed Extension Instruction */                \
-            DISPATCH_RV32C()                                      \
-        }                                                         \
-    }
-
-#define EXEC(instr)                          \
-    {                                        \
-        /* dispatch this opcode */           \
-        if (unlikely(!op_##instr(rv, insn))) \
-            return;                          \
-        /* increment the cycles csr */       \
-        rv->csr_cycle++;                     \
-    }
-
-#define TARGET(instr)         \
-    op_##instr : EXEC(instr); \
-    DISPATCH();
-
-    DISPATCH();
-
-    /* main loop */
-    TARGET(load)
-    TARGET(op_imm)
-    TARGET(auipc)
-    TARGET(store)
-    TARGET(op)
-    TARGET(lui)
-    TARGET(branch)
-    TARGET(jalr)
-    TARGET(jal)
-    TARGET(system)
-#if RV32_HAS(EXT_C)
-    TARGET(caddi4spn)
-    TARGET(caddi)
-    TARGET(cslli)
-    TARGET(cjal)
-    TARGET(clw)
-    TARGET(cli)
-    TARGET(clwsp)
-    TARGET(clui)
-    TARGET(cmisc_alu)
-    TARGET(ccr)
-    TARGET(cj)
-    TARGET(csw)
-    TARGET(cbeqz)
-    TARGET(cswsp)
-    TARGET(cbnez)
-#endif
+    opcode_t opcode_handler_entry[] = {
```

Comment:
Don't do that! You should introduce IR while decoupling RISC-V instruction decoder from `emulate.c`. And then, execute the opcode in the form of predefined IR.

Reply:
Understand! Thanks for your reminder.

Reply:
On the other hand, create function `rv_emulate` to execute all implemented instructions as following:
```c
static bool rv_emulate(struct riscv_t *rv, struct rv_insn_t *ir)
{
    switch (ir->opcode) {
    /* RV32I Base Instruction Set*/
    case rv_insn_lui:
    case rv_insn_auipc:
    case rv_insn_jal:
    case rv_insn_jalr:
    case rv_insn_beq:
    case rv_insn_bne:
    case rv_insn_blt:
    ...
}
```

Reply:
`rv_emulate` can be effectively implemented with computed-goto as what `fastrv32` does.